### ISSUE TITLE
Something is annoyingly wrong with query execution at the toplevel

### ISSUE THREAD

**Issue by jamescheney** [links-lang/links#435]
Consider the following (assuming a database configured as in `examples/relational_lenses`):

```
links> var db = database "links";
db = (database postgresql:links:localhost:5432:...  Database
links> var albumsTable =
...... table "albums"
...... with (album: String, quantity: Int)
...... tablekeys [["album"]]
...... from db;
albumsTable = (table albums) : TableHandle((album:String,quantity:Int),(album:String,quantity:Int),(album:String,quantity:Int))
links> for (a <-- albumsTable) [a];
*** Fatal error : Function definition in query: 2070
links> fun foo () {for (a <-- albumsTable) [a]};
foo = fun : () -> [(album:String,quantity:Int)]
links> foo()
...... ;
[(album = "Disintegration", quantity = 7), (album = "Wish", quantity = 5),
 (album = "Paris", quantity = 4), (album = "Galore", quantity = 1),
 (album = "Show", quantity = 3)] : [(album:String,quantity:Int)]
links> 
```

That is, running the trivial query `for (a <-- albumsTable) [a]` (as well as more interesting queries) fails at the toplevel (also if we write `query{}` around the expression).  But wrapping it inside a function and then calling the function succeeds.  The problem also appears if you run the whole thing as a standalone links file:

```
var db = database "links";

var albumsTable =
	table "albums"
	with (album: String, quantity: Int)
	tablekeys [["album"]]
	from db;

for (t <-- albumsTable) [t]
```

This is in links 0.8 - haven't tested on master yet.  This definitely used to work, and is somewhat annoying from the point of view of demoing.

**Comment by jstolarek** [comment]
`git bisect` says the culprit is:

```
d93415a039a11a8d02a91cb94a3b6108c30bf231 is the first bad commit
commit d93415a039a11a8d02a91cb94a3b6108c30bf231
Author: Frank Emrich <10780413+frank-emrich@users.noreply.github.com>
Date:   Fri Jun 29 17:54:53 2018 +0100

    Towards de-cycling recursive types + type-checking the IR (#355)

    * Created a visitor for data types

    * Moved IR visitors from ir.ml to irTraversals.ml
    Created "backend.ml", which contains backend processing pipeline

    * Moved type visitor upwards in Types.ml to make it usuable for more functions in there. Implemented refreshing of recursive variables when transforming recursive types/rows

    * Removed some of the irTraversals, re-enabled the remaining ones
```

@frank-emrich, leaving this with you.

**Comment by jamescheney** [comment]
Here is a smaller (and non database dependent) example, showing the resulting IR:

```
links> query{for (x <- []) []};
compiling to IR
compiled IR
e: ([`Fun (((2139, ((_::Any) {}~> [(|_::(Any,Base))], "_fun__g158", `Local)),
         ([], [(2138, (_::Any, "", `Local))],
          ([`Let (((2136, ([_::Any], "_for__g157", `Local)),
                   ([], `Return (`Variable (2138)))));
             `Let (((2137, (_::Any, "x", `Local)),
                    ([], `Return (`Variable (2138)))))
             ],
           `Return (`TApp ((`Variable (55), [(|_::(Any,Base))]))))),
         None, `Unknown))
   ],
 `Apply ((`TApp ((`Variable (234), [_::Any; { wild }; (|_::(Any,Base))])),
          [`Variable (2139); `TApp ((`Variable (55), [_::Any]))])))
*** Fatal error : Function definition in query: 2139
```

It looks like, sometime after closure conversion, `for (x <- e1) e2` is getting desugared to `concatMap(\x. e1, e2)`.

However, if the query block is inside a function:
```
links> fun foo() {query {for (x <- []) []} };
compiling to IR
compiled IR
([`Fun (((2208, (() -> [(|_::(Any,Base))], "foo", `Global)),
         ([a::Row(Any,Any); a::Row(Any,Base)], [],
          ([],
           `Special (`Query ((None,
                              ([`Fun (((2207,
                                        ((_::Any) -> [(|_::(Any,Base))],
                                         "_fun__g190", `Local)),
                                       ([], [(2206, (_::Any, "", `Local))],
                                        ([`Let (((2204,
                                                  ([_::Any], "_for__g189",
                                                   `Local)),
                                                 ([],
                                                  `Return (`Variable (2206)))));
                                           `Let (((2205,
                                                   (_::Any, "x", `Local)),
                                                  ([],
                                                   `Return (`Variable (2206)))))
                                           ],
                                         `Return (`TApp ((`Variable (
                                                          55),
                                                          [(|_::(Any,Base))]))))),
                                       None, `Unknown))
                                 ],
                               `Apply ((`TApp ((`Variable (234),
                                                [_::Any; { |_::Any };
                                                  (|_::(Any,Base))])),
                                        [`Variable (2207);
                                          `TApp ((`Variable (55), [_::Any]))]))),
                              [(|_::(Any,Base))]))))),
         None, `Unknown))
   ],
 `Return (`Extend (({}, None))))
Finishing process MAIN
foo = fun : () -> [(|_::(Any,Base))]
```
then it looks like hte same code is generated, but when we apply the function:
```
links> foo();
compiling to IR
compiled IR
([],
 `Apply ((`TApp ((`TAbs (([a::Row(Any,Any); a::Row(Any,Base)],
                          `TApp ((`Variable (2208),
                                  [{ |_::Any }; { |_::(Any,Base) }])))),
                  [{ |wild }; { |_::(Any,Base) }])),
          [])))
e: ([],
 `Apply ((`TApp ((`Variable (234), [_::Any; { |_::Any }; (|_::(Any,Base))])),
          [`Closure ((2207, [{ |_::(Any,Base) }], `Extend (({}, None))));
            `TApp ((`Variable (55), [_::Any]))])))
v: `Concat ([])
Finishing process MAIN
[] : [(|_::(Any,Base))]
```

somehow we get a closure in the right place?  (The "e:" and "v:" messages are a debugging print messages in query.ml that are normally commented out that I've enabled to try to see what's going on).

**Comment by frank-emrich** [comment]
Sorry, I didn't see Janek's bisecting result earlier. I'll have a look at this.

**Comment by frank-emrich** [comment]
Fixed in 8257f6a1884c20f917f4903fa02b6a08725a7632.

This was a rather embarassing bug: We would run each IR program through the backend piepline (including closure conversion), but then we would give the original program to the evaluator, instead of the transformed one. However, the backend pipeline adds all functions (after closure conversion) to a global, mutable hash table. This is why we didn't notice the prolem earlier: Whenever the evaluator looked up a function, we would get correctly closure-converted code, only non-function expression were broken.

**Comment by jamescheney** [comment]
Thanks!
